File: MinStack.js
class MinStack {
  constructor(stack = [], min = Number.MAX_SAFE_INTEGER) {
    this.stack = stack;
    this.min = min;
  }

  push(num) {
    this.min = num < this.min ? num : this.min;

    this.stack.push(num);
  }

  pop() {
    let popped = -1;

    if (this.stack.length > 0) {
      popped = this.stack.pop();

      if (popped === this.min) {
        this.min = Math.min(...this.stack);
      }
    }
    return popped;
  }

  top() {
    const top = this.stack.pop();

    this.stack.push(top);

    return top;
  }

  getMin() {
    return this.min;
  }
}

// ------------ *** ------------

const stack = new MinStack();

stack.push(-2);
stack.push(0);
stack.push(-3);
console.log(stack.getMin());
console.log(stack.pop());
console.log(stack.top());
console.log(stack.getMin());

File: min_stack.py
""" design a stack DSA """


def push(stack, min_stack, num):
    if not stack:
        stack.append(num)
        min_stack.append(num)
        return

    smallest = min_stack[-1]
    min_stack.append(min(smallest, num))
    stack.append(num)


def pop(stack, min_stack):
    if not stack:
        return

    popped = stack.pop()
    _ = min_stack.pop()

    return popped


def get_min(min_stack):
    if not min_stack:
        return

    return min_stack[-1]


stack = []
min_stack = []

while True:
    command = input()

    match command:
        case "pop":
            print(pop(stack, min_stack))

        case "push":
            push(stack, min_stack, input())

        case "getMin":
            print(get_min(min_stack))

        case "Minstack":
            stack = []
            min_stack = []

        case _:
            break  # return
